lib/TPP/ConvertPerfToFunc.cpp

```diff
@@ -0,0 +1,361 @@
+        .Case<MemRefType>([&](Type t) {
+          auto memrefType = cast<MemRefType>(t);
+          auto unrankedMemref = UnrankedMemRefType::get(
+              memrefType.getElementType(), memrefType.getMemorySpace());
+          results.push_back(unrankedMemref);
+        })
+        .Case<TensorType>([&](Type t) {
+          auto tensorType = cast<TensorType>(t);
+          auto unrankedTensor =
+              UnrankedTensorType::get(tensorType.getElementType());
+          results.push_back(unrankedTensor);
+        })
+        .Case<TimerType>([&](Type t) {
+          auto i64 = IntegerType::get(b.getContext(), 64);
+          results.push_back(i64);
+        })
+        .Default([&](Type t) { results.push_back(t); });
+  }
+
+  return results;
+}
+
+// Similar to 'extractNormalizedTypes' but inserts conversions from original
+// types to normalized ones when possible.
+// The conversions aim to simplify runtime function signature generation by, for
+// example, erasing explicit memref and tensor shapes.
+static SmallVector<Value> getNormalizedOperands(OpBuilder &b, Location loc,
+                                                ValueRange operands) {
+  SmallVector<Value> res;
+  res.reserve(operands.size());
+
+  for (Value op : operands) {
+    auto type = extractNormalizedTypes(b, op);
+    TypeSwitch<Type>(op.getType())
+        .Case<MemRefType>([&](Type t) {
+          Value cast = b.create<memref::CastOp>(loc, type, op);
+          res.push_back(cast);
+        })
+        .Case<TensorType>([&](Type t) {
+          Value cast = b.create<tensor::CastOp>(loc, type, op);
+          res.push_back(cast);
+        })
+        .Default([&](Type t) { res.push_back(op); });
+  }
+
+  return res;
+}
+
+// Apply custom function name mangling for various data types.
+// It is assumed that all relevant perf operations accept only unranked memory
+// types. This allows for simpler name mangling and leaner perf runtime.
+static void applyTypeMangling(std::string &name, Type type) {
+  llvm::raw_string_ostream mangledName(name);
+
+  TypeSwitch<Type>(type)
+      .Case<MemRefType>([&](Type t) {
+        mangledName << "_memref"
+                    << "_" << cast<MemRefType>(t).getElementType();
+      })
+      .Case<TensorType>([&](Type t) {
+        mangledName << "_tensor"
+                    << "_" << cast<TensorType>(t).getElementType();
+      })
+      .Default([&](Type t) { mangledName << "_" << t; });
+}
+
+// Generate function implementation for perf.mean operation.
+static void buildPerfMeanFunc(Location loc, func::FuncOp func, Operation *op,
+                              PatternRewriter &rewriter) {
+  OpBuilder::InsertionGuard guard(rewriter);
+
+  // Create function body
+  Block *block = func.addEntryBlock();
+  rewriter.setInsertionPointToEnd(block);
+
+  // Check assumptions on function arguments.
+  auto argTypes = func.getFunctionType().getInputs();
+  assert((argTypes.size() == 1) && "expect only 1 function argument");
```

Comment:
These asserts all disappear when you build the function in one go offline and use it later on the rewriter for the perf op.

Reply:
Good point, moved to proper op errors.